client-runtime/src/main/java/com/microsoft/rest/v2/http/RxNettyAdapter.java

```diff
@@ -42,52 +47,91 @@ private SSLEngine getSSLEngine(String host) {
-        URI uri;
+        Single<? extends HttpResponse> result;
+
         try {
-            uri = new URI(request.url());
-        } catch (URISyntaxException e) {
-            return Single.error(e);
-        }
+            final URI uri = new URI(request.url());
 
-        Map<String, Set<Object>> rxnHeaders = new HashMap<>();
-        for (HttpHeader header : request.headers()) {
-            rxnHeaders.put(header.name(), Collections.<Object>singleton(header.value()));
-        }
+            Map<String, Set<Object>> rxnHeaders = new HashMap<>();
+            for (HttpHeader header : request.headers()) {
+                rxnHeaders.put(header.name(), Collections.<Object>singleton(header.value()));
+            }
 
-        String mimeType = request.mimeType();
-        if (mimeType != null) {
-            rxnHeaders.put("Content-Type", Collections.<Object>singleton(mimeType));
-        }
+            String mimeType = request.mimeType();
+            if (mimeType != null) {
+                rxnHeaders.put("Content-Type", Collections.<Object>singleton(mimeType));
+            }
 
-        String body = request.body();
-        if (body != null) {
-            rxnHeaders.put("Content-Length", Collections.<Object>singleton(String.valueOf(body.length())));
-        }
+            boolean isSecure = "https".equalsIgnoreCase(uri.getScheme());
+            io.reactivex.netty.protocol.http.client.HttpClient<ByteBuf, ByteBuf> rxnClient =
+                    io.reactivex.netty.protocol.http.client.HttpClient.newClient(uri.getHost(), isSecure ? 443 : 80);
+
+            if (isSecure) {
+                rxnClient = rxnClient.secure(getSSLEngine(uri.getHost()));
+            }
+
+            HttpClientRequest<ByteBuf, ByteBuf> rxnReq = rxnClient
+                    .createRequest(HttpMethod.valueOf(request.httpMethod()), uri.toASCIIString())
+                    .addHeaders(rxnHeaders);
+
+            Observable<HttpClientResponse<ByteBuf>> obsResponse = rxnReq;
 
-        boolean isSecure = "https".equalsIgnoreCase(uri.getScheme());
-        io.reactivex.netty.protocol.http.client.HttpClient<ByteBuf, ByteBuf> rxnClient =
-                io.reactivex.netty.protocol.http.client.HttpClient.newClient(uri.getHost(), isSecure ? 443 : 80);
+            final InputStream body = request.body();
+            if (body != null) {
+                obsResponse = rxnReq.writeBytesContent(toByteArrayObservable(body));
+            }
 
-        if (isSecure) {
-            rxnClient = rxnClient.secure(getSSLEngine(uri.getHost()));
+            result = obsResponse
+                    .map(new Func1<HttpClientResponse<ByteBuf>, HttpResponse>() {
+                        @Override
+                        public HttpResponse call(HttpClientResponse<ByteBuf> rxnRes) {
+                            return new RxNettyResponse(rxnRes);
+                        }
+                    })
+                    .toSingle();
+        } catch (URISyntaxException e) {
```

Comment:
Why expand the try block to include a bunch of stuff that won't throw that exception? Do you prefer to avoid early returns?

Reply:
I do generally not like early returns. I find it easier to understand code if it has one entry and one exit. In this case, I really wanted to combine the Single.error() call so that any exception would go to the same place and would skip the rest of the logic inside the try...catch block.